.buildkite/steps/test-demo-notebooks.sh

```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+
+set -xeo pipefail
+
+stellargraph_dir="$PWD"
+
+SPLIT="${BUILDKITE_PARALLEL_JOB_COUNT:-1}"
+INDEX="${BUILDKITE_PARALLEL_JOB:-0}"
+
+echo "--- :books: collecting notebooks"
+# Create array with all notebooks in demo directories.
+cd "${stellargraph_dir}"
+NOTEBOOKS=()
+while IFS= read -r -d $'\0'; do
+  NOTEBOOKS+=("$REPLY")
+done < <(find "${stellargraph_dir}/demos" -name "*.ipynb" -print0)
+
+NUM_NOTEBOOKS_TOTAL=${#NOTEBOOKS[@]}
+
+if [ "$NUM_NOTEBOOKS_TOTAL" -ne "$SPLIT" ]; then
+  echo "Buildkite step parallelism is set to $SPLIT but found $NUM_NOTEBOOKS_TOTAL notebooks. Please update parallelism to match the number of notebooks!"
+  exit 1
+fi
+
+echo "--- installing dependencies"
+pip install -q --no-cache-dir '.[demos]'
+
+echo "--- listing dependency versions"
+pip freeze
+
+f=${NOTEBOOKS[$INDEX]}
+
+echo "+++ :python: running $f"
+case $(basename "$f") in
+  'attacks_clustering_analysis.ipynb' | 'hateful-twitters-interpretability.ipynb' | 'hateful-twitters.ipynb' | 'stellargraph-attri2vec-DBLP.ipynb' | \
+    'node-link-importance-demo-gat.ipynb' | 'node-link-importance-demo-gcn.ipynb' | 'node-link-importance-demo-gcn-sparse.ipynb' | 'rgcn-aifb-node-classification-example.ipynb' | \
+    'stellargraph-metapath2vec.ipynb' | \
+    'ensemble-node-classification-example.ipynb' | 'cora-gcn-links-example.ipynb')
+    # FIXME: these notebooks do not yet work on CI (#818 - datasets can't be downloaded)
+    # FIXME: these notebooks do not yet work on CI (#819 - out-of-memory)
+    # FIXME: these notebooks do not yet work on CI (#833 - too slow)
+    # FIXME: these notebooks do not yet work on CI (#816 - missing dataset download)
+    exit 2 # this will be a soft-fail for buildkite
+    ;;
+  *) # fine, run this one
+    cd "$(dirname "$f")"
+    jupyter nbconvert --to notebook --execute --ExecutePreprocessor.timeout=2400 "$f"
```

Comment:
Could you say why you went with this approach rather than the `papermill` one that `neon` uses?

I think we should strongly consider that one, or at least cribbing more of its features:

- `papermill` seems to give better command line output about the cells as they fail (in particular, better stdout and stderr handling)
- `papermill` will more easily generalise to handling #819, #820 and #833 if we need to reduce epochs or similar with its parameterisation functionality
- `neon` uploads the notebooks as buildkite artefacts so it's easy to download them and inspect the full output with colour etc.

Reply:
no good reason: I just started with nbconvert from my outdated neon checkout - have updated with papermill and artefact upload from neon (the concept of behaviour changes is good)